### PR TITLE
Config: Can add static headers to email messages

### PR DESCRIPTION
**Pull request by owensmallwood**
fixes https://github.com/grafana/hosted-grafana/issues/5058

You can add static headers config in the `smtp.static_headers` section for headers to get added to emails.

Example grafana config could be:
```
#################################### SMTP / Emailing #####################
[smtp.static_headers]
Foo=value1
Bar=value2
```

**Example**
Add the following to your grafana config:
```
[smtp.static_headers]
Foo-Header=foo_val
Barheader=abc123
``` 

Edit: The header validation was updated to only work with smtp headers before I recorded this 👇 
Then send an email and inspect the headers:
smtp_static_headers

### CODE REVIEW

File: pkg/services/notifications/notifications.go
Comment (owensmallwood) [inline pkg/services/notifications/notifications.go:200]: These are the only two spots where `buildEmailMessage()` gets called. Would be a good place to add tenant-level custom headers like slug.

**Comment by owensmallwood** [comment]
@hairyhenderson Tested this locally with the smtprelay logging and it works as expected

File: pkg/setting/setting_smtp.go
Comment (hairyhenderson) [inline pkg/setting/setting_smtp.go:66]: nit:

```suggestion
			return fmt.Errorf("invalid label name %q in [smtp.static_headers] configuration", labelName)
```

File: pkg/services/notifications/smtp.go
Comment (alexweav) [inline pkg/services/notifications/smtp.go:69]: I wonder if we should reverse the order here, and apply the static headers before the From/To/Subject ones?

Only rationale being, it opens up some odd paths like overriding `From` globally to something unusual. It seems like we'd always want to respect the individual message over the static headers in this case.

Since this all comes from the config.ini anyway, and isn't UI-editable, it's probably not a big deal, so consider this a nitpick if we disagree.

File: pkg/setting/setting_smtp.go
Comment (hairyhenderson) [inline pkg/setting/setting_smtp.go:70]: nit:

```suggestion
			return fmt.Errorf("invalid label value for name %q in [smtp.static_headers] configuration: %s", labelName, labelValue)
```

File: pkg/setting/setting_smtp.go
Comment (alexweav) [inline pkg/setting/setting_smtp.go:65]: Is there an SMTP-specific validation we could use? For example, dots are allowed in SMTP headers, but not in prometheus labels.

From a search it looks like SMTP uses standard MIME headers (though I am not an expert in this area, so please double check), and there's a tool in the stdlib for dealing with these: https://pkg.go.dev/mime#WordDecoder

To validate, we might be able to just encode and decode the given header, and pass if it works with no errors.

**Review by alexweav — COMMENTED** [review]
Overall this is looking great! I'll defer to the codeowners of the SMTP service for final approval, but this seems good to me. I had just a couple minor nitpicks.

File: pkg/services/notifications/smtp.go
Comment (hairyhenderson) [reply pkg/services/notifications/smtp.go:69]: > Only rationale being, it opens up some odd paths like overriding `From` globally to something unusual.

That makes a lot of sense IMO - we should definitely disallow that.

Comment (owensmallwood) [reply pkg/services/notifications/smtp.go:69]: Yeah, may as well change the order. Doesn't hurt 👍

File: pkg/setting/setting_smtp.go
Comment (owensmallwood) [reply pkg/setting/setting_smtp.go:65]: Sure, I'll add some regex to check for that

**Review by undef1nd — APPROVED** [review]

File: pkg/setting/setting_smtp.go
Comment (hairyhenderson) [inline pkg/setting/setting_smtp.go:30]: _technically_ these are "mail headers" - they're not part of the SMTP protocol (they're also not strictly MIME headers, as they don't relate to MIME directly, and are not part of multipart messages)

```suggestion
// validates mail headers
var mailHeaderRegex = regexp.MustCompile(`^[A-Z][A-Za-z0-9]*(-[A-Z][A-Za-z0-9]*)*$`)
```

Comment (hairyhenderson) [reply pkg/setting/setting_smtp.go:30]: I should also say, my (incredibly pedantic) suggestion is mainly intended for internal variable naming. I think it still makes sense to name the config section `smtp.static_headers`, for example, because it only relates to messages sent over SMTP 😅

**Review by zserge — APPROVED** [review]

File: pkg/setting/setting_smtp.go
Comment (hairyhenderson) [inline pkg/setting/setting_smtp.go:70]: I think the error message could provide a better hint to how to fix the issue:

```suggestion
			return fmt.Errorf("header %q in [smtp.static_headers] configuration: must follow canonical MIME form", key.Name())
```

**Review by chri2547 — APPROVED** [review]

**Review by hairyhenderson — APPROVED** [review]
LGTM! Thanks for working on this @owensmallwood 🥳

**Pull request merged**
